src/Processors/QueryPlan/BufferChunksTransform.cpp

```diff
@@ -0,0 +1,84 @@
+#include <Processors/QueryPlan/BufferChunksTransform.h>
+
+namespace DB
+{
+
+BufferChunksTransform::BufferChunksTransform(
+    const Block & header_,
+    size_t max_rows_to_buffer_,
+    size_t max_bytes_to_buffer_,
+    size_t limit_)
+    : IProcessor({header_}, {header_})
+    , input(inputs.front())
+    , output(outputs.front())
+    , max_rows_to_buffer(max_rows_to_buffer_)
+    , max_bytes_to_buffer(max_bytes_to_buffer_)
+    , limit(limit_)
+{
+}
+
+IProcessor::Status BufferChunksTransform::prepare()
+{
+    if (output.isFinished())
+    {
+        chunks = {};
+        input.close();
+        return Status::Finished;
+    }
+
+    if (output.canPush())
+    {
+        input.setNeeded();
+
+        if (!chunks.empty())
+        {
+            auto chunk = std::move(chunks.front());
+            chunks.pop();
+
+            num_buffered_rows -= chunk.getNumRows();
+            num_buffered_bytes -= chunk.bytes();
+
+            output.push(std::move(chunk));
+        }
+        else if (input.hasData())
+        {
+            auto chunk = pullChunk();
+            output.push(std::move(chunk));
+        }
+        else if (input.isFinished())
+        {
+            output.finish();
+            return Status::Finished;
+        }
```

Comment:
I'm not sure if any problem could actually happen with the current implementation, but just for a piece of mind I'd move this `if` outside of the `if (output.canPush())` condition just right on the l.28

Reply:
Ok, probably makes sense.